services/webnn/ort/graph_builder_ort.cc

```diff
@@ -258,6 +259,658 @@ GraphBuilderOrt::GraphBuilderOrt(
+        //              |
+        //           reshape_1
+        //              |
+        //  input_a    transpose
+        //     \       /
+        //        matmul
+        if (input_id_to_pre_matmul_transpose_reshape_dq_map.contains(
+                output_id)) {
+          // Add reshape_2
+          graph_fusion_info.fusible_operations_set.insert(operation.get());
+          graph_fusion_info.output_id_to_fusible_matmulnbits_map[output_id] =
+              operation.get();
+          // Add dequantizeLinear
+          const mojom::Operation* dq =
+              input_id_to_pre_matmul_transpose_reshape_dq_map[output_id];
+          graph_fusion_info.fusible_operations_set.insert(dq);
+          graph_fusion_info.output_id_to_fusible_matmulnbits_map
+              [dq->get_dequantize_linear()->output_operand_id] = dq;
+          // Add reshape_1
+          const mojom::Operation* reshape =
+              input_id_to_pre_matmul_transpose_reshape_map
+                  [dq->get_dequantize_linear()->output_operand_id];
+          graph_fusion_info.fusible_operations_set.insert(reshape);
+          graph_fusion_info.output_id_to_fusible_matmulnbits_map
+              [reshape->get_reshape()->output_operand_id] = reshape;
+          // Add transpose
+          const mojom::Operation* transpose =
+              input_id_to_pre_matmul_transpose_map[reshape->get_reshape()
+                                                       ->output_operand_id];
+          graph_fusion_info.fusible_operations_set.insert(transpose);
+          graph_fusion_info.output_id_to_fusible_matmulnbits_map
+              [transpose->get_transpose()->output_operand_id] = transpose;
+        }
+
+        break;
+      }
+      case mojom::Operation::Tag::kDequantizeLinear: {
+        CHECK_EQ(operation_connectivity.output_ids.size(), 1U);
+        uint64_t output_id = operation_connectivity.output_ids[0];
+        if (!input_id_to_pre_matmul_transpose_reshape_map.contains(output_id) ||
+            operand_id_to_use_count_map[output_id] != 1) {
+          break;
+        }
+        const mojom::DequantizeLinearPtr& dequantize_linear =
+            operation->get_dequantize_linear();
+        // Input, zero point must be constants.
+        if (!graph_info.constant_operand_ids_to_handles.contains(
+                dequantize_linear->input_operand_id) ||
+            !graph_info.constant_operand_ids_to_handles.contains(
+                dequantize_linear->zero_point_operand_id)) {
+          break;
+        }
+        // If scale is a constant, the reshape_2 will be folded into scale by
+        // constant_folding in WebNN blink side. The fusion patter will looks
+        // like this:
+        //     input   scale   zeroPoint
+        //         \     |      /
+        //         dequantizeLinear
+        //              |
+        //           reshape
+        //              |
+        //  inputa    transpose
+        //     \       /
+        //        matmul
```

Comment:
Is this the common case? Should we focus on this pattern?

Reply:
After sync with @Honry, we suppose scale always be constant. Remove another pattern.